include/pulsar/c/consumer_configuration.h

```diff
@@ -98,6 +98,21 @@ typedef struct {
     long timeoutMs;
 } pulsar_consumer_batch_receive_policy_t;
 
+typedef struct {
+    // Name of the dead topic where the failing messages are sent.
+    // The default value is: sourceTopicName + "-" + subscriptionName + "-DLQ"
+    const char *dead_letter_topic;
+    // Maximum number of times that a message is redelivered before being sent to the dead letter queue.
+    // - The maxRedeliverCount must be greater than 0.
+    // - The default value is {INT_MAX} (DLQ is not enabled)
+    const int max_redeliver_count;
+    // Name of the initial subscription name of the dead letter topic.
+    // If this field is not set, the initial subscription for the dead letter topic is not created.
+    // If this field is set but the broker's `allowAutoSubscriptionCreation` is disabled, the DLQ producer
+    // fails to be created.
+    const char *initial_subscription_name;
+} pulsar_consumer_config_dead_letter_policy_t;
```

Comment:
In addition, I think we'd better not to use raw struct. In C, the initial field values of a struct is not determined. For example,

```c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
  int x;
  const char *s;
} A;

int main(int argc, char *argv[]) {
  A a;
  printf("%d | %s\n", a.x, a.s);
  return 0;
}
```

```bash
$ gcc 0.c -O0
$ ./a.out
891409248 | (null)
$ gcc 0.c -O2
$ ./a.out
0 | (null)
```

Unlike C++ (from 11), a struct could not have initial values. 

In addition, we'd better not to return a struct because its size could be greater than 8 bytes. POSIX APIs usually add a pointer argument as the output argument. Take [setrlimit/getrlimit](https://linux.die.net/man/2/setrlimit) for example:

```c
int getrlimit(int resource, struct rlimit *rlim);
int setrlimit(int resource, const struct rlimit *rlim);

struct rlimit {
    rlim_t rlim_cur;  /* Soft limit */
    rlim_t rlim_max;  /* Hard limit (ceiling for rlim_cur) */
};
```

These problems exist for `pulsar_consumer_batch_receive_policy_t` as well. I will push a PR to fix it.

Reply:
Regarding the default value, I just thought again and now I think it's okay to use raw struct types but we must document it well, especially for the **default values**. For example,

```c
typedef struct {
    // ...
    // If it's null, use sourceTopicName + "-" + subscriptionName + "-DLQ" as the value
    const char *dead_letter_topic;
    // ...
    // If it's not greater than 0, treat it as INT_MAX
    const int max_redeliver_count;
    // ...
    // If it's null, the initial subscription for the dead letter topic is not created.
    const char *initial_subscription_name;
} pulsar_consumer_config_dead_letter_policy_t;
```

i.e. a `NULL` string and a zero integer are treated as special default values. And the caller should set all these fields explicitly.

Reply:
> In addition, I think we'd better not to use raw struct. In C, the initial field values of a struct is not determined. For example

emm, I agree with your case. But, what is a better example of this case?

Reply:
> i.e. a NULL string and a zero integer are treated as special default values. And the caller should set all these fields explicitly.

I agree, Users need to set these fields, and we do not promise the behavior of not setting these fields, which should be the consensus of C programming.

Reply:
> But, what is a better example of this case?

I checked some POSIX API designs. There are two typical ways.

1. Require users to set all fields explicitly.
2. Use the getter to initialize the current struct.

The `setrlimit`/`getrlimit` pair is an example.

```c
// Set two customized values
struct rlimit rl;
rl.rlim_cur = 100;
rl.rlim_max = 1024;
setrlimit(RLIMIT_XXX, &rl);
```

or

```c
// Get the default values first, then only modify one field
struct rlimit rl;
getrlimit(RLIMIT_XXX, &rl);
rl.rlim_max = 1024;
setrlimit(RLIMIT_XXX, &rl);
```

Reply:
If you prefer the OOP style, you can also add a function like:

```c
struct A {
    int x;
    const char* s;
};

void init_a(struct A* a) {
    a.x = 0;
    a.s = NULL;
}
```

Then it requires users to use `init_a` to initialize struct `A`. The `init_a` function acts like C++'s constructor.

BTW, I think all of them are okay.